plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -712,6 +748,60 @@ private void finishOptimize(IcebergTableExecuteHandle executeHandle, Collection<
         transaction = null;
     }
 
+    @Override
+    public void executeTableExecute(ConnectorSession session, ConnectorTableExecuteHandle tableExecuteHandle)
+    {
+        IcebergTableExecuteHandle executeHandle = (IcebergTableExecuteHandle) tableExecuteHandle;
+        switch (executeHandle.getProcedureId()) {
+            case VACUUM:
+                executeVacuum(session, executeHandle);
+                return;
+            default:
+                throw new IllegalArgumentException("Unknown procedure '" + executeHandle.getProcedureId() + "'");
+        }
+    }
+
+    private void executeVacuum(ConnectorSession session, IcebergTableExecuteHandle executeHandle)
+    {
+        IcebergVacuumHandle vacuumHandle = (IcebergVacuumHandle) executeHandle.getProcedureHandle();
+        Table table = catalog.loadTable(session, executeHandle.getSchemaTableName());
+        long retentionThreshold = vacuumHandle.getRetentionThreshold().toMillis();
+        long expireTimestampMillis = System.currentTimeMillis() - retentionThreshold;
+        table.expireSnapshots()
+                .expireOlderThan(expireTimestampMillis)
+                .cleanExpiredFiles(true)
+                .commit();
+        deleteOrphanFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
+    }
+
+    private void deleteOrphanFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        Set<String> validDataFilePaths = Streams.stream(table.snapshots())
+                .map(snapshot -> snapshot.snapshotId())
+                .flatMap(snapshotId -> Streams.stream(table.newScan().useSnapshot(snapshotId).planFiles()))
+                .map(fileScanTask -> fileScanTask.file().path().toString())
+                .collect(toImmutableSet());
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(new HdfsEnvironment.HdfsContext(session), new Path(table.location()));
+            RemoteIterator<LocatedFileStatus> allDataFiles = fileSystem.listFiles(new Path(table.location() + "/data/"), true);
```

Comment:
The data files aren't necessarily in this directory, or even necessarily in the same file system as the table location. Take a look at the `LocationProvider` 

I think you need a check similar to https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java#L365

Reply:
The data files that are "somewhere else" won't be cleaned up. But i agree a check like that could be nice.
We can fail for now, and revisit later.

Reply:
It may make sense to consider different catalog, however the reasoning is related to the spark architecture, the way it works. They don't want to do too much on spark driver which actually makes sense, here we explicitly want to do everything on the coordinator and not in a distributed manner so I would not worry about that.

Reply:
I added vacuuming metadata